nodego/supervisor.go

```diff
@@ -0,0 +1,255 @@
+	}
+
+	if loggingContext.currentLogBatch == nil {
+		return os.Stderr.Write(entry.consoleOutput(loggingContext.functionExecutionID))
+	}
+
+	// Start a new batch if the current one would grow too much.
+	if len(loggingContext.currentLogBatch.Entries)+1 > maxLogBatchEntries || loggingContext.currentLogBatch.payloadLength+len(entry.TextPayload) > maxLogBatchLength {
+		startNewLogBatch()
+	}
+
+	loggingContext.currentLogBatch.addEntry(entry)
+	return len(entry.TextPayload), nil
+}
+
+func (batch *logBatch) report() error {
+	if len(batch.Entries) == 0 {
+		return nil
+	}
+
+	if err := postToSupervisor("/_ah/log", batch, supervisorLogTimeout); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func flushLogBatches() {
+	select {
+	case logBatch := <-logBatchesToReport:
+		go flushLogBatches()
+		if err := logBatch.report(); err != nil {
+			fmt.Fprintln(os.Stderr, err.Error())
+			killInstance()
+		}
+	default:
+		startNewLogBatch()
+	}
+}
+
+func logBatchReportWorker() {
+	for logBatch := range logBatchesToReport {
+		// TODO it may be nice to have a time base buffering method on top of this.
+		<-logBatch.ready
+
+		loggingContext.Lock()
+		if logBatch == loggingContext.currentLogBatch {
+			startNewLogBatch()
+		}
+		loggingContext.Unlock()
+
+		if err := logBatch.report(); err != nil {
+			fmt.Fprintln(os.Stderr, err.Error())
+			killInstance()
+		}
+	}
+}
+
+func newSupervisorRequest(path string, v interface{}) (*http.Request, error) {
+	postData, err := json.Marshal(v)
+	if err != nil {
+		return nil, err
+	}
+
+	req, err := http.NewRequest("POST", (&url.URL{
+		Scheme: "http",
+		Host:   supervisorHostname + ":" + supervisorInternalPort,
+		Path:   path,
+	}).String(), bytes.NewBuffer(postData))
+	if err != nil {
+		return nil, err
+	}
+
+	req.Header.Add("Content-Type", "application/json")
+	req.Header.Add("Content-Length", strconv.Itoa(len(postData)))
+
+	return req, nil
+}
+
+func postToSupervisor(path string, v interface{}, timeout time.Duration) error {
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+
+	req, err := newSupervisorRequest(path, v)
+	if err != nil {
+		return err
+	}
+
+	resp, err := ctxhttp.Do(ctx, nil, req)
+	if err == ctx.Err() {
+		return errors.New("timeout when calling supervisor")
+	} else if err != nil {
+		return fmt.Errorf("error when calling supervisor: %s\n\n%s\n", err.Error(), debug.Stack())
+	}
+
+	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
+		err = fmt.Errorf("incorrect response code from supervisor: %d\n", resp.StatusCode)
+	}
+
+	return err
+}
+
+func killInstance() {
+	postToSupervisor("/_ah/kill", nil, supervisorKillTimeout)
+	os.Exit(16)
+}
+
+func loggerMiddleware(handler http.Handler) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		loggingContext.Lock()
+		loggingContext.functionExecutionID = r.Header.Get("Function-Execution-Id")
+		loggingContext.Unlock()
+
+		defer func() {
+			go flushLogBatches()
```

Comment:
Ideally all log lines tagged with a specific execution ID should be flushed. This can be guaranteed by moving this line below the resetting of loggingContext.functionExecutionID.